File: packages/next/src/executor.ts
import { Core, executorSymbol, Meta } from "./types";
import type { Escapable } from "./helpers";

function createExecutor<T>(
  factory: Core.NoDependencyFn<T> | Core.DependentFn<T, unknown>,
  dependencies:
    | undefined
    | Core.UExecutor
    | Array<Core.UExecutor>
    | Record<string, Core.UExecutor>,
  metas: Meta.Meta[] | undefined
): Core.Executor<T> {
  const executor = {
    [executorSymbol]: "main",
    factory: (_: unknown, controller: Core.Controller) => {
      if (dependencies === undefined) {
        const f = factory as Core.NoDependencyFn<T>;
        return f(controller);
      }

      const f = factory as Core.DependentFn<T, unknown>;
      return f(_, controller);
    },
    dependencies,
    metas: metas,
  } as unknown as Core.Executor<T>;

  const lazyExecutor = {
    [executorSymbol]: "lazy",
    dependencies: undefined,
    executor,
    factory: undefined,
    metas: metas,
  } satisfies Core.Lazy<T>;

  const reactiveExecutor = {
    [executorSymbol]: "reactive",
    executor,
    factory: undefined,
    dependencies: undefined,
    metas: metas,
  } satisfies Core.Reactive<T>;

  const staticExecutor = {
    [executorSymbol]: "static",
    dependencies: undefined,
    factory: undefined,
    metas: metas,
    executor,
  } satisfies Core.Static<T>;

  Object.defineProperties(executor, {
    lazy: {
      value: lazyExecutor,
      writable: false,
      configurable: false,
      enumerable: false,
    },
    reactive: {
      value: reactiveExecutor,
      writable: false,
      configurable: false,
      enumerable: false,
    },
    static: {
      value: staticExecutor,
      writable: false,
      configurable: false,
      enumerable: false,
    },
  });

  return executor;
}

export function isLazyExecutor(
  executor: Core.BaseExecutor<unknown>
): executor is Core.Lazy<unknown> {
  return executor[executorSymbol] === "lazy";
}

export function isReactiveExecutor(
  executor: Core.BaseExecutor<unknown>
): executor is Core.Reactive<unknown> {
  return executor[executorSymbol] === "reactive";
}

export function isStaticExecutor(
  executor: Core.BaseExecutor<unknown>
): executor is Core.Static<unknown> {
  return executor[executorSymbol] === "static";
}

export function isMainExecutor(
  executor: unknown
): executor is Core.Executor<unknown> {
  return isExecutor(executor) && executor[executorSymbol] === "main";
}

export function isExecutor<T>(input: unknown): input is Core.BaseExecutor<T> {
  return typeof input === "object" && input !== null && executorSymbol in input;
}

export function provide<T>(
  factory: Core.NoDependencyFn<T>,
  ...metas: Meta.Meta[]
): Core.Executor<T> {
  return createExecutor(factory, undefined, metas);
}

export function derive<T, D extends Core.BaseExecutor<unknown>>(
  dependencies: D,
  factory: Core.DependentFn<T, Core.InferOutput<D>>,
  ...metas: Meta.Meta[]
): Core.Executor<T>;

export function derive<
  T,
  D extends
    | ReadonlyArray<Core.BaseExecutor<unknown>>
    | Record<string, Core.BaseExecutor<unknown>>
>(
  dependencies: { [K in keyof D]: D[K] },
  factory: Core.DependentFn<T, { [K in keyof D]: Core.InferOutput<D[K]> }>,
  ...metas: Meta.Meta[]
): Core.Executor<T>;

export function derive<T, D>(
  pdependencies:
    | Core.BaseExecutor<D>
    | Array<Core.BaseExecutor<D>>
    | Record<string, Core.BaseExecutor<unknown>>,
  pfactory:
    | Core.DependentFn<T, Core.InferOutput<D>>
    | Core.DependentFn<T, { [K in keyof D]: Core.InferOutput<D[K]> }>,
  ...metas: Meta.Meta[]
): Core.Executor<T> {
  return createExecutor(pfactory as any, pdependencies, metas);
}

export function preset<T>(
  e: Core.Executor<T> | Escapable<T>,
  v: T
): Core.Preset<T> {
  const executor = isExecutor(e) ? e : e.escape();

  return {
    [executorSymbol]: "preset",
    value: v,
    executor,
  };
}

export function placeholder<V>(...metas: Meta.Meta<unknown>[]): Core.Executor<V> {
  return provide<V>(() => {
    throw new Error("Placeholder executor cannot be resolved");
  }, ...metas);
}

File: pkg/core/executor.go
package core

import (
	"context"
	"fmt"
	"reflect"
)

// executorBase is the base implementation for all executors
type executorBase[T any] struct {
	kind         ExecutorKind
	dependencies []any
	metadata     map[string]any
}

// Kind returns the kind of executor
func (e *executorBase[T]) Kind() ExecutorKind {
	return e.kind
}

// Dependencies returns the dependencies of this executor
func (e *executorBase[T]) Dependencies() []any {
	return e.dependencies
}

// Metadata returns the metadata associated with this executor
func (e *executorBase[T]) Metadata() map[string]any {
	return e.metadata
}

// WithMeta adds metadata to the executor
func (e *executorBase[T]) WithMeta(key string, value any) Executor[T] {
	if e.metadata == nil {
		e.metadata = make(map[string]any)
	}
	e.metadata[key] = value
	return e
}

// mainExecutor is the primary executor implementation
type mainExecutor[T any, D any] struct {
	executorBase[T]
	factory func(D, Controller) (T, error)
}

// Lazy returns a lazy version of this executor
func (e *mainExecutor[T, D]) Lazy() Executor[Accessor[T]] {
	return &lazyExecutor[T]{
		executorBase: executorBase[Accessor[T]]{
			kind:         KindLazy,
			dependencies: nil,
			metadata:     e.metadata,
		},
		main: e,
	}
}

// Reactive returns a reactive version of this executor
func (e *mainExecutor[T, D]) Reactive() Executor[T] {
	return &reactiveExecutor[T]{
		executorBase: executorBase[T]{
			kind:         KindReactive,
			dependencies: nil,
			metadata:     e.metadata,
		},
		main: e,
	}
}

// Static returns a static version of this executor
func (e *mainExecutor[T, D]) Static() Executor[Accessor[T]] {
	return &staticExecutor[T]{
		executorBase: executorBase[Accessor[T]]{
			kind:         KindStatic,
			dependencies: nil,
			metadata:     e.metadata,
		},
		main: e,
	}
}

// lazyExecutor is an executor that doesn't resolve immediately
type lazyExecutor[T any] struct {
	executorBase[Accessor[T]]
	main *mainExecutor[T, any]
}

// reactiveExecutor is an executor that updates when dependencies change
type reactiveExecutor[T any] struct {
	executorBase[T]
	main *mainExecutor[T, any]
}

// staticExecutor is an executor that doesn't track for reactivity
type staticExecutor[T any] struct {
	executorBase[Accessor[T]]
	main *mainExecutor[T, any]
}

// Provide creates an executor with no dependencies
func Provide[T any](factory func(Controller) (T, error)) MainExecutor[T] {
	return &mainExecutor[T, struct{}]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: nil,
			metadata:     make(map[string]any),
		},
		factory: func(_ struct{}, ctrl Controller) (T, error) {
			return factory(ctrl)
		},
	}
}

// Derive creates an executor with a single dependency
func Derive[T any, D any](
	dependency Executor[D],
	factory func(D, Controller) (T, error),
) MainExecutor[T] {
	return &mainExecutor[T, D]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: []any{dependency},
			metadata:     make(map[string]any),
		},
		factory: factory,
	}
}

// DeriveMulti creates an executor with multiple dependencies of the same type
func DeriveMulti[T any, D any](
	dependencies []Executor[D],
	factory func([]D, Controller) (T, error),
) MainExecutor[T] {
	deps := make([]any, len(dependencies))
	for i, dep := range dependencies {
		deps[i] = dep
	}

	return &mainExecutor[T, []D]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: deps,
			metadata:     make(map[string]any),
		},
		factory: factory,
	}
}

// DeriveMap creates an executor with map dependencies
func DeriveMap[T any, K comparable, V any](
	dependencies map[K]Executor[V],
	factory func(map[K]V, Controller) (T, error),
) MainExecutor[T] {
	deps := make([]any, 0, len(dependencies))
	for _, dep := range dependencies {
		deps = append(deps, dep)
	}

	return &mainExecutor[T, map[K]V]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: deps,
			metadata:     make(map[string]any),
		},
		factory: factory,
	}
}

// DeriveTyped creates an executor with multiple dependencies of different types
func DeriveTyped[T any, D1 any, D2 any](
	dep1 Executor[D1],
	dep2 Executor[D2],
	factory func(D1, D2, Controller) (T, error),
) MainExecutor[T] {
	return &mainExecutor[T, struct {
		Dep1 D1
		Dep2 D2
	}]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: []any{dep1, dep2},
			metadata:     make(map[string]any),
		},
		factory: func(deps struct {
			Dep1 D1
			Dep2 D2
		}, ctrl Controller) (T, error) {
			return factory(deps.Dep1, deps.Dep2, ctrl)
		},
	}
}

// DeriveTyped3 creates an executor with three dependencies of different types
func DeriveTyped3[T any, D1 any, D2 any, D3 any](
	dep1 Executor[D1],
	dep2 Executor[D2],
	dep3 Executor[D3],
	factory func(D1, D2, D3, Controller) (T, error),
) MainExecutor[T] {
	return &mainExecutor[T, struct {
		Dep1 D1
		Dep2 D2
		Dep3 D3
	}]{
		executorBase: executorBase[T]{
			kind:         KindMain,
			dependencies: []any{dep1, dep2, dep3},
			metadata:     make(map[string]any),
		},
		factory: func(deps struct {
			Dep1 D1
			Dep2 D2
			Dep3 D3
		}, ctrl Controller) (T, error) {
			return factory(deps.Dep1, deps.Dep2, deps.Dep3, ctrl)
		},
	}
}

// IsExecutor checks if a value is an executor
func IsExecutor(value any) bool {
	if value == nil {
		return false
	}

	t := reflect.TypeOf(value)
	if t.Kind() != reflect.Ptr {
		return false
	}

	// Check if it implements the Executor interface
	execType := reflect.TypeOf((*Executor[any])(nil)).Elem()
	return reflect.PtrTo(t.Elem()).Implements(execType)
}

// IsMainExecutor checks if a value is a main executor
func IsMainExecutor(value any) bool {
	if !IsExecutor(value) {
		return false
	}

	// Use type assertion to check the kind
	if exec, ok := value.(interface{ Kind() ExecutorKind }); ok {
		return exec.Kind() == KindMain
	}
	return false
}

// IsLazyExecutor checks if a value is a lazy executor
func IsLazyExecutor(value any) bool {
	if !IsExecutor(value) {
		return false
	}

	// Use type assertion to check the kind
	if exec, ok := value.(interface{ Kind() ExecutorKind }); ok {
		return exec.Kind() == KindLazy
	}
	return false
}

// IsReactiveExecutor checks if a value is a reactive executor
func IsReactiveExecutor(value any) bool {
	if !IsExecutor(value) {
		return false
	}

	// Use type assertion to check the kind
	if exec, ok := value.(interface{ Kind() ExecutorKind }); ok {
		return exec.Kind() == KindReactive
	}
	return false
}

// IsStaticExecutor checks if a value is a static executor
func IsStaticExecutor(value any) bool {
	if !IsExecutor(value) {
		return false
	}

	// Use type assertion to check the kind
	if exec, ok := value.(interface{ Kind() ExecutorKind }); ok {
		return exec.Kind() == KindStatic
	}
	return false
}

// CreatePreset creates a preset value for an executor
func CreatePreset[T any](executor Executor[T], value T) Preset[T] {
	return Preset[T]{
		Executor: executor,
		Value:    value,
	}
}

